src/interfaces/signatures.cpp

```diff
@@ -0,0 +1,263 @@
+#include "signatures.hpp"
+#include <cassert>
+
+namespace
+{
+    std::tuple<uint32_t, span<uint8_t>> read_compressed_uint(span<uint8_t> signature)
+    {
+        ULONG value;
+        signature = slice(signature, CorSigUncompressData(signature, &value));
+        return std::make_tuple(value, signature);
+    }
+
+    std::tuple<int32_t, span<uint8_t>> read_compressed_int(span<uint8_t> signature)
+    {
+        int value;
+        signature = slice(signature, CorSigUncompressSignedInt(signature, &value));
+        return std::make_tuple(value, signature);
+    }
+
+    std::tuple<mdToken, span<uint8_t>> read_compressed_token(span<uint8_t> signature)
+    {
+        mdToken value;
+        signature = slice(signature, CorSigUncompressToken(signature, &value));
+        return std::make_tuple(value, signature);
+    }
+
+    class SignatureElementCallback
+    {
+    public:
+        virtual void OnReadByte(uint8_t byte) = 0;
+        virtual void OnReadCompressedUInt(uint32_t value) = 0;
+        virtual void OnReadCompressedInt(int32_t value) = 0;
+        virtual void OnReadToken(mdToken token) = 0;
+    };
+
+    span<uint8_t> WalkSignatureElement(span<uint8_t> signature, SignatureElementCallback& callback)
```

Comment:
Instead of this approach, I would convert to a template and perhaps use some kind of meta programming to restrict use.

Reply:
Concepts would have been useful here... sigh.

Reply:
I have an idea of a way to model this as a lambda. I'll try it out. C++14 would be nice here, but I think I can make my idea work in C++11.